Add Sortablejs as element (dragable objects!)

EmberLightVFX opened the pull request:
This PR adds Sortable to NiceGUI! ([](https://sortablejs.github.io/Sortable/))
I used the original sortable.js and not any vue adaptations as those were a bit old and were made out of multiple files so I have no idea how to implement them.

Any way, I think I have gotten everything to work correctly. I copied Sortable's examples into the NiceGUI sortable example so you can try out all the different options you can use.

Currently the example contains a bunch of custom css. It would be best to add them into the sortable.py but I would like to know what the best way would be to implement it?

evnchn commented:
> not any vue adaptations as those were a bit old

Wise move. https://github.com/SortableJS/vue.draggable.next is 4 years old. 

To clarify, I think your file `sortable.complete.esm.js` is the same one as in https://www.npmjs.com/package/sortablejs?activeTab=code, right? 

This is because, eventually, we would need a script to fetch it automatically (check out `npm.py` and `npm.json`)

> Currently the example contains a bunch of custom css. It would be best to add them into the sortable.py

Almost. If you have a good bit of reasonable default for styles, they would commonly go to https://github.com/zauberzeug/nicegui/blob/main/nicegui/static/nicegui.css

Typically, styles are applied to elements with class `nicegui-*`. A bunch of other elements got the treatment there. 

It would be great to include just the mininal style so that the sortable doesnt look like hot garbage, but not anything else on top. After all, the user may also want to apply their own styles as well. 

> Any way, I think I have gotten everything to work correctly. 

I'm fine to take your word for it now, but eventually there should be tests. Not sure how we may "drag" stuff with ChromeDriver, though. 

---

@falkoschindler correct me if I am wrong.

EmberLightVFX commented:
Updated with some missing examples and options for sortablejs.
I also added some default css in the css file :)
The only example not currenty working is the Thresholds one. Need to take a better look at it...

About the npm,py/json I would gladly take some help with that, or some instructions on how that part of NiceGUI works.
[https://www.npmjs.com/package/sortablejs?activeTab=code](https://www.npmjs.com/package/sortablejs?activeTab=code) is the correct one. I'm only using the file **sortable.complete.esm.js** that's under /sortable/modular/sortable.complete.esm.js as it contains all official plugins for sortable.

EmberLightVFX commented:
I have now fixed the threshold example to fully work and also made it possible to drag cards into other cards in the nested example.

The only thing left is the npm part and I would love some help with that as I have no idea how you would set it up correctly.

falkoschindler commented:
Thanks for the pull request, @EmberLightVFX! Bringing SortableJS to NiceGUI would be an awesome feature!
I'll review your code as soon as possible. The update of npm.json shouldn't be a problem. I can do that. 👍🏻

evnchn commented:
@EmberLightVFX Include the following in your `npm.json`:

```json
  "sortable": {
    "package": "sortablejs",
    "destination": "nicegui/elements/lib",
    "keep": [
      "package/modular/sortable.complete.esm.js"
    ],
    "rename": {
      "package/modular/": ""
    },
    "version": "1.15.6"
  }
```

Hopefully, this should download `sortable.complete.esm.js` and place the file at `nicegui/elements/lib/sortable/sortable.complete.esm.js`

falkoschindler reviewed (changes requested):
Hi @EmberLightVFX, thanks for this massive pull request!

I roughly reviewed your code and played around with `ui.sortable` for a bit. Here are some of my findings:

- With 1600 lines of NiceGUI code (not including 3300 lines of SortableJS) this PR is huge and hard to review in its current form.
- In contrast to other elements like `ui.aggrid`, `ui.echart`, or `ui.leaflet`, the new `ui.sortable` doesn't simply expect an `options` dictionary, but provides explicit parameters for every single option. Even though this might feel nice from the user perspective, it makes the element very hard to maintain. Every upstream change can break our API, requiring extensive regression tests and potentially slowing down major releases like 3.0. Generic options, on the other hand, reduce complexity on our side while providing full flexibility for the user.
- Alternatively, if you'd be willing to invest that much time and devotion into this element, you could split it into a separate repository and PyPI package. This would clearly communicate that the NiceGUI core development team isn't responsible for maintaining `ui.sortable`. (It's not my prefered option though.)
- We should move the example main.py into the documentation. By splitting it into small, independent, interactive demos they become more accessible and reproducible.
- At the moment there are no pytests at all. We should at least test basic functionality to ensure the library is loaded and sortable content is shown on the screen. Testing drag and drop is challenging, but we should cover it in some basic test.
- While I'm arguing for a more generic implementation, it would be great to have special event arguments including the most relevant information.
- I'm wondering if we could render `ui.sortable` with flex layout by default, like most of NiceGUI's layout elements (`ui.row`, `ui.column`, `ui.card`, `ui.scroll_area`, ...). This would allow you to nest some cards inside `ui.sortable` and it just looks nice.
- On the other hand, when trying to reproduce lists with list items, you might want to nest `ui.item` inside `ui.list` and somehow combine it with `ui.sortable`. This brings me to maybe the most challenging idea: Is "sortable" really a UI element, or a property of _any_ existing container element? I mean, maybe we should reconsider the API and check if we could create something like `with ui.row().sortable(): ui.label(A); ...`.
- I noticed some confusing animation in certain situations, which also happens with the original online examples. So I hope SortableJS is stable enough and won't cause tons of bug reports on our end. (At the time of writing, they have 468 unresolved issues...)

Please don't get me wrong: I love what you did so far and the sheer amount of code is impressive. But before merging anything on main, we have to challenge our decisions. Once a feature is out in the wild, it's very hard to change or take back anything. Allowing to make containers sortable is a very high priority feature and we definitely need something like this in NiceGUI. We just have to make sure to do it right.

Let's start with the most fundamental questions:

- Is SortableJS stable enough to be included in NiceGUI? (I think so.)
- Should this PR and its code remain in this repo? (I do hope so.)
- Could we turn it into a feature of any `ui.element` like `column.sortable()`?

When these questions are answered, other tasks like moving examples, writing tests, simplifying the parameter list, adding special event arguments etc. will be relatively straight forward.

evnchn commented:
First off, if not for @falkoschindler mentioning, I would not have noticed the sheer scale of this PR. @EmberLightVFX it is impressive work. Be proud of yourselves and have some rest if need be. 

---

But, @falkoschindler I need to address **one important point:** 

> Could we turn it into a feature of any `ui.element` like `column.sortable()`

As you have mentioned, SortableJS may not be the most stable, and we may or may not have to limit SortableJS children elements to an allowlist (can't imagine what sorting Leaflet maps or Plotly diagrams would be like 🤯). 

Turning it into a feature of any `ui.element`, with your wishful programming example of `with ui.row().sortable():`, means we are getting our feets **deep into the water**, which may not be a good idea. 

At least certainly not a good idea now. Maybe later, when a good chunk of people have used SortableJS with NiceGUI and find it stable and generic enough to work with a bunch of elements. But, we'll have to re-review by then. 

---

With that said, I think the following action items in the following order makes the most sense: 

- [ ] Let `ui.sortable` simply expect an `options` dictionary to decouple itself from upstream parameter definition changes, and simplify the NiceGUI implementation by a whole lot. 
  - Remeber NiceGUI is all for [a shallow layer around existing libraries](https://github.com/zauberzeug/nicegui/discussions/4734#discussioncomment-13126503). 
  - This can make the PR more reviewable by reducing line count
- [ ] Move the examples in `examples/sortable/main.py` into separate ones in documentation. 
  - This can make the PR more reviewable by visual confirmation
  - We can also report more explicitly, say, demo 6 is broken, and we only need to look at demo 6 which is isolated. 
  - Also helps with getting new users to use this element. 
- [ ] Adopt `flex` layout if possible
  - Helps make it easy to assess whether it is possible to leverage the entire / most of NiceGUI elements with `ui.sortable`
  - ["flex layout by default" is in our website front page](https://nicegui.io/#:~:text=flex%20layout%20by%20default)
- [ ] Pytest test cases
  - It's super useful since sometimes we break JS libraries when we mess with the underlying infrastructure. Saved us in #4539 or else we would have shipped broken Mermaid. 

Not saying the PR is necessarily "done" whe all these action items are done, but it'll be closer than ever, and we can spot the remaining issues much more easily by then.

EmberLightVFX commented:
Finally I got some time over to work on this PR again.

Some updates from the feedback!
* Use options dict like aggrid, leaflet etc to match all other elements
* Split up the example into documentation
* Write some pytests
* Use flex layout by default

I added a drag-and-drop function to NiceGUIs screen testing as selenium have that functionality internally. Maybe it should be its own PR but it's only 6 lines of code and (currently) only sortable that uses it, so I added it to this PR.


* `While I'm arguing for a more generic implementation, it would be great to have special event arguments including the most relevant information.`
I did not fully understand what you're after here. Could you explain in another way for me?

-----

If there is anything more that needs to be fixed with this PR, just say and I'll do my best to do it as fast as possible :)

evnchn commented:
Wow. You sure are efficient! Now I see some potential improvements to the codebase. I am not sure if I can collaborate but I will put them out here first. 

- [ ] Can you do `dependencies=['lib/sortable/sortable.complete.esm.js']` instead of `self.add_resource(Path(__file__).parent / 'lib' / 'sortable')`, then you can `import { default as Sortable } from 'sortable';` in `sortable.js` and change dynamic import to static import, and avoid defining `this.Sortable`. 

<img width="871" alt="{0C4A73BC-9FFA-486E-859D-43B956510D20}" src="https://github.com/user-attachments/assets/027d3cf6-1923-4ef5-9b84-17fdf3763f22" />

<img width="870" alt="{D1191EB8-65EF-4E98-A45B-95FB0ADB4329}" src="https://github.com/user-attachments/assets/04f1bae8-901f-4053-88e8-873e68f48f48" />

- [ ] If you already did `default_classes='nicegui-sortable'`, I don't think we need `self.classes('nicegui-sortable')`? 

- [ ] `on_end` to `on_deselect` callbacks, should be `Handler` of `EventArguments` or maybe even `ValueChangeEventArguments`. Do a code search on `Handler[ValueChangeEventArguments]` and you'll instantly get what I mean. 
  - [ ] Then `e` will not be `Any` anymore <br> <img width="316" alt="{7FFE60E3-8ADB-4B29-8A59-5433E5B6F4FA}" src="https://github.com/user-attachments/assets/9d21c82c-ee5b-4c36-83f1-33db07ccc884" />

---

That's what I can spot so far. There may be more, but it this _should_ be enough to keep you busy for a short while.

evnchn commented:
idk if it works but I put down all my changes so far in https://github.com/EmberLightVFX/nicegui/pull/1

evnchn commented:
- [ ] Enable retroactively add handler & mutiple handler, by means of following:
  - Code search `tree.py` and it will give you some inspirations. 
  - ```py
    self._select_handlers = [on_select] if on_select else []
    ```
  - ```py
    def on_select(self, callback: Handler[ValueChangeEventArguments]) -> Self:
        """Add a callback to be invoked when the selection changes."""
        self._props.setdefault('selected', None)
        self._select_handlers.append(callback)
        return self
    ```
  - ```py
        def handle_selected(e: GenericEventArguments) -> None:
            update_prop('selected', e.args)
            for handler in self._select_handlers:
                handle_event(handler, ValueChangeEventArguments(sender=self, client=self.client, value=e.args))
        self.on('update:selected', handle_selected)
    ```

EmberLightVFX commented:
Thank you @evnchn for all the notes and the PR! I merged it and fixed some small stuff around the code.

I didn't really understand your comment https://github.com/zauberzeug/nicegui/pull/4656#issuecomment-2903289729 about retroactively add handler & mutiple handler so I would love to get that cleared a bit.

Thanks!

evnchn commented:
@EmberLightVFX _Generally speaking_ a NiceGUI element accepts both passing in the change handler on construction, and retroactively. 

Consider `ui.tree`

<img width="234" alt="{75FAAF3C-DB97-49EA-9693-F7D688DAF63C}" src="https://github.com/user-attachments/assets/97402f4c-095e-4bbc-aa1a-5a5ce3a1c6d1" />

So I wonder if you can make it the case too?

evnchn commented:
https://github.com/EmberLightVFX/nicegui/pull/2

@EmberLightVFX don't merge right away since idk if this is right or not

@falkoschindler can you see if I am messing things up or making things better? Thanks.

EmberLightVFX commented:
@evnchn I'm actually working on that right now. I have done exactly like you except I made a generic handle function and passed the correct handler list to it to reduce the amount of code.

I'm currently fighting with pytest after updating this as I'm getting `Message: timeout: Timed out receiving message from renderer: 3.734` for the last tests if I run all the tests. Induvidual runs passes... Does pytest pass fully for you?

evnchn commented:
Sorry. As I have said I'm developing on Windows, and per test case I only have 50% success rate (WSL does NOT help), making it infeasible to run all tests locally. 

I'm waiting for the GitHub actions result to see how much, if any, did I mess up the tests.

EmberLightVFX commented:
Oh really? I'm also running on Windows but I have 100% successfully run all the tests. I'm running locally and not any docker over WSL.

I'm using UV so I init my NiceGUI environment like this:
```bash
uv venv -p 3.8

.venv/Scrupts/activate

uv pip install Poetry

Poetry install
```

evnchn commented:
Tests are green in my attempt: https://github.com/evnchn/nicegui/actions/runs/15212866217

So idk what problem you are facing...

falkoschindler commented:
> can you see if I am messing things up or making things better? Thanks.

@evnchn Looks good. But it's really a lot of events. Do we really need all of them? Maybe we can reduce them to the most common ones and keep the rest of them available via the generic `.on()`?

EmberLightVFX commented:
The current state of this code is a bit of a mess.
I'm currently fighting the true clone example. It works mostly but some times the cloning breaks and what ever I do I'm only getting `DOMException: Node.insertBefore: Child to insert before is not a child of this node` errors.

Checking the content of the python object and the DOM object I can see they aren't matching but I don't understand why...

I have also added all missing events like onUnchoose etc. When I have gotten the clone problem fixed we could scale down the amount of events to maybe 1-2 or something.

I added a debug example to the docs where you can see what event gets triggered when. 

So in short: The code is now a bit of a mess while trying to figure out what's going on. After that the code will get cleaned up.

evnchn commented:
Add oil! Like idk why this is marked with a major. I hope this doesn't stress you out... Love to see the final result 💪

falkoschindler commented:
Oh, the "major" tag says: "Important, but not urgent". In other words: It would be a really useful feature we'd love to add to NiceGUI, but it's not a critical bug that needs fixing as soon as possible.

Regarding the list of events we need to find a sweet spot. Take `ui.input` for example: It would be a pain to use the generic `.on('update:model-value', ...)` and extract the `value` from GenericEventArguments every time. On the other hand, the user rarely needs the events "blur", "focus" or "clear", so the generic API is a good trade-off.

And maybe we don't need cloning in the very first version of `ui.sortable`. As long as we don't make bad API choices that might prevent us from adding it later, we _can_ decide to focus on the core functionality for now.

evnchn commented:
In principle it should, since there's multiple sortable in the demo page I'd assume. 

Also if not, I'd consider it a bug, and we need to fix it. Basically all UI elements (short of header, footer, left right drawer) can have more than one of each kind, after all.

TM0088 commented:
> It works with any amount of sortable lists!
The example is amazing, most of the features are working great.

I have found that 'show_current_order' cannot get the correct result right after adding item. Properly because item is added as the 1st item by using nicegui 'with' function while item is added to the last in sortablejs logic.

Besides, i can not get 'getSelected', 'select' and 'deselect' working properly in multi-drag sortable object, is it still in process or may you provide some example?

jochso commented:
Great feature! Would it be possible to add expandable/collapsible lists as well? I think that is a feature NiceGui would really benefit from 👍

TM0088 commented:
> Update time!
> 1. Changed the init kwargs to the ones I wrote in my eariler post: https://github.com/zauberzeug/nicegui/pull/4656#issuecomment-2907518159
> 2. I removed the getSelected function as the selection would get de-selected if you pressed a button or something like that. It's also not part of native SortableJS.
> 3. Fixed the code for select/deselect so it now works as expected.
> 
> I tried to find the bug @TM0088 talked about with show_current_order but I couldn't replicate it.
> Could you record a video showing how you get the bug? And if you're using custom code, please share it also.
> 
> If there isn't any show_current_order bug then I would say this PR is ready for its last review and merge!

Is there any workaround for getSelected function for multi-drag objects?

evnchn commented:
https://github.com/zauberzeug/nicegui/pull/4656/commits/a2aea470f97b23b9b952ccc0c4800e35211348a2: Yes but then our JS is then not updated by any script anymore. Moving forward full migration to #5021's approach is the way to go eventually.

EmberLightVFX commented:
Thanks @evnchn !
I was actually about to ask about this.

Why removing the npn.json became it's own commit was because something happened in the merge and I ended up with an empty npm file but got said no files has bin changed, so I had to manually remove it. No idea why that happened 🤷

evnchn commented:
You can check #5021 and what happened there. But in short t's gone and move into each element.

falkoschindler commented:
@EmberLightVFX I was just about to finally review this PR, when I saw your merge commit. Just to be sure: Are you currently working on it? We probably need to update the ESM dependency management to see if everything works with NiceGUI 3.

EmberLightVFX commented:
I were suppose to update my drop_zone branch but accidentally updated this one 😅 I'm not working on this one.
I would love some help on how to correctly set up the esm dependencie for NiceGUI 3!

falkoschindler commented:
Hi everyone,

Let me give a short update on what I've been thinking about the last few weeks.

### Which library

SortableJS is probably good enough for now and we don't need to change it. If we find a good abstraction and the user doesn't have to deal with implementation details, we can always replace the underlying JavaScript library with something else in the future.

### API design

I still don't like the fact that the current implementation requires the user to deal with generic events and map IDs to actual data items. PR #5464 tried to simplify things and gathered valuable [feedback](https://github.com/zauberzeug/nicegui/pull/5464#issuecomment-3556601262).

Right now I'm leaning towards a `.sortable()` method that can be called from container elements like `ui.column`, `ui.row`, `ui.card`, `ui.list`, and a `.sortable_item()` method to mark children as draggable and to attach data:
```py
with ui.column().sortable() as column:
    ui.label('A').sortable_item(data='A')
    ui.label('B').sortable_item(data='B')
    with ui.label('C').sortable_item(data='C', handle='.handle'):
        ui.icon('drag_handle').classes('handle')
```

The `.sortable` namespace can also hold additional properties and methods, e.g. to activate and deactivate the sortable feature:
```py
ui.button('Activate', on_click=column.sortable.activate)
ui.button('Deactivate', on_click=column.sortable.deactivate)
```

The `.sortable()` method can accept an event handler which can access the current data list of the child elements.

### Beta namespace

@evnchn proposed a `ui.beta` namespace to hold experimental features. I'm not so sure about it because NiceGUI could more and more become a collection of unfinished features. Instead I think it would be better to release elements like `ui.sortable` as standalone packages so that users can try them out and provide feedback.

I'm planning to continue working on this feature in January. But of course, I can't promise any release date.

EmberLightVFX commented:
I'm super glad to see conversations about this! I feel I'm a bit low on energy on this PR and I don't have much time to put into it for at least a couple of months so I'm more than happy to give this PR over to @phifuh if you're up for it?

evnchn commented:
I still haven't got myself into this PR, but: 

- If your JS lags hard enough that you see `reloading because handshake failed for clientID xxxx` then you may have blocked the JS main thread long enough that Socket.IO connection cannot be sustained. While #5784 will salvage the situation, it is not ideal. 
- Nothing comes to mind for big changes in 3.2 through 3.6 which could break your `ReportValidationStorage`. I would recommend that you can bisect by yourself to land on a specific PR which broke things.

falkoschindler commented:
Thanks once again for this PR, @EmberLightVFX — it kicked off a great discussion and helped shape our thinking on how to best integrate SortableJS into NiceGUI.

After extensive discussion in #932, we concluded that "sortable" is better modeled as a _feature on existing containers_ rather than a standalone element. A `ui.column` that can be reordered is still a column. PR #5855 implements this approach via a `.make_sortable()` method on container elements like `ui.column`, `ui.row`, `ui.card`, etc.

Closing this in favor of #5855. Your contribution and the ideas from this PR were instrumental in getting us there — thank you!
